lib/Runtime/Types/SimpleTypeHandler.cpp

```diff
@@ -956,7 +973,7 @@ namespace Js
     template<size_t size>
     BOOL SimpleTypeHandler<size>::SetAccessors(DynamicObject* instance, PropertyId propertyId, Var getter, Var setter, PropertyOperationFlags flags)
     {
-        if (DoConvertToPathType(instance->GetDynamicType()))
+        if (DoConvertToPathType(instance->GetDynamicType()) && !HasAnyPropertyDeleted())
```

Comment:
There are other usages of `DoConvertToPathPath` (see `SimpleTypeHandler::AddProperty`). Should they also be checking for deleted properties? If so, would it make sense to move this "has deleted properties" check into that function?

Reply:
There are other places where DoConvertToPathType is used, but they are used inside GetIsLocked() check. And it seems, from the comment I read in SetAttribute, that GetIsLocked returns false if any property got deleted.
However I meant to put that in AddProperty - thanks for catching that.

Reply:
This is a weaker solution than the one we discussed offline. Converting to a new type handler is equivalent to evolving the type from scratch. In this case it needs to be able to handle any situation that simple types can handle and path types can't. Deleted properties may not be the only one. We need an API that tries to convert to path type but may return some kind of dictionary type handler.

Reply:
But why create path type (and spent unnecessary cycles to create and throw away) when we are sure we are going to create dictionary type? Right now I only know for deleted property.

Reply:
also on other places we don't follow the same rule
```
        if (propertyCount >= sizeof(descriptors)/sizeof(SimplePropertyDescriptor))
        {
            Assert(propertyId != Constants::NoProperty);
            if (DoConvertToPathType(instance->GetDynamicType()))
            {
                return ConvertToPathType(instance)->SetPropertyWithAttributes(instance, propertyId, value, attributes, info, flags);
            }
            else
            {
                PropertyRecord const* propertyRecord = scriptContext->GetPropertyName(propertyId);
                return ConvertToSimpleDictionaryType(instance)->AddProperty(instance, propertyRecord, value, attributes, info, flags, possibleSideEffects);
            }
        }
```
we decide based on certain condition that we need to create path type or dictionary type.